Material type: Fresh-frozen or FFPE source - in a corresponding RNA manifest

Tumor percentage in tissue source (% malignant cells) - should probably be in a corresponding RNA manifest?
Extraction sample (regular cut, macrodissection, etc.) - should probably be in a corresponding RNA manifest?

Pre-sequencing QC method/instrument - ? 
Pre-sequencing RNA QC metrics: 
	- RIN
	- 260/280
	- 260/230
	- DV200
	- RQS
	- Volume of RNA (ng)
	- Concentration of RNA (ng/uL)

May include: Total RNA input for library prep (in ng)

Library construction protocol/kit: Yes

Library QC method/instrument: Yes

Library QC metrics: 
	- Pre-hybridization library size (number of bp)
	- Pre-hybridization library quantity (ng)
	- Final library size (number of bp)
	- Final library concentration (mM, ng/uL, or CT value from qPCR)
	- Target mean length (number of bp): insert length?
	- Sequencing platform: Yes

Minimum depth of sequencing: number of paired-end (PE) reads: *Needs clarification*, This is used as a lab cut-off for failed sequencing

Minimum depth of sequencing: number of total reads: Need clarification, For Illumina, this is double of Minimum depth of sequencing: number of paired-end (PE) reads:

Multiplexing used or not: Yes

Sequencing QC method/package (e.g., RSeQC): Yes (site-specific pipeline)

Sequencing QC metrics: Yes
	- % stranded
	- % mRNA bases
	- % rRNA bases
	- % aligned
	- Target depth number of reads or paired reads
	- Insert size
	- Duplication rate : needs clarification
	- 3’, 5’ Bias (0-1)  

Post-sequencing QC metrics: Yes
	- Median transcript integrity number (medTIN)
	- Number of reads
	- Number samples failed/per batch
	- Definition of failed: which QC metrics are used for pass/fail decisions

Reference genome (e.g., hg38)

 

